R/sequential.R

```diff
@@ -0,0 +1,94 @@
+#' Perform multiple requests in sequence
+#'
+#' Given a list of requests, this function performs each in turn, returning
+#' a list of responses. It's slower than [req_perform_parallel()] but
+#' has fewer limitations.
+#'
+#' @inheritParams req_perform_parallel
+#' @inheritParams req_perform_iteratively
+#' @param on_error What should happen if one of the requests throws an
+#'   HTTP error?
+#'
+#'   * `stop`, the default: stop iterating and throw an error
+#'   * `return`: stop iterating and return all the successful responses so far.
+#'   * `ignore`: continue iterating, recording errors in the result.
+#' @export
+#' @return
+#' A list, the same length as `reqs`.
+#'
+#' If `error_on` is `"return"` and it errors on the ith request, the ith
+#' element of the result will be an error object, and the remaining elements
+#' will be `NULL`.
+#'
+#' If `error_on` is `"ignore"`, it will be a mix of requests and errors.
+#' @examples
+#' # req_perform_sequential() if the API allows you to request data for
+#' # multiple objects in each request, but in total you still have more
+#' # objects than can fit in on request.
+#'
+#' req <- request("https://api.restful-api.dev/objects")
+#'
+#' # Imagine we have 50 ids:
+#' ids <- sort(sample(100, 50))
+#'
+#' # But the API only allows us to request 10 at time. So we first use split
+#' # and some modulo arithmetic magic to generate chunks of length 10
+#' chunks <- unname(split(ids, (seq_along(ids) - 1) %/% 10))
+#'
+#' # Then we use lapply to generate one request for each chunk:
+#' reqs <- chunks %>% lapply(\(idx) req %>% req_url_query(id = idx, .multi = "comma"))
+#'
+#' # Then we can perform them all and get the results
+#' \dontrun{
+#' resps <- reqs %>% req_perform_sequential()
+#' resps_data(resps, \(resp) resp_body_json(resp))
+#' }
+req_perform_sequential <- function(reqs,
+                                   paths = NULL,
+                                   on_error = c("stop", "return", "ignore"),
+                                   progress = TRUE) {
+  if (!is_bare_list(reqs)) {
+    stop_input_type(reqs, "a list")
+  }
+  if (!is.null(paths)) {
+    if (length(reqs) != length(paths)) {
```

Comment:
Maybe also check the type with `check_character()`? I also opened an issue for `req_perform()` #364.